# Diabetes 130

In [2]:
import zipfile
import pandas as pd
from anndata import AnnData

In [3]:
pd.set_option("display.max_columns", None)

## Getting the dataset

In [ ]:
!wget http://archive.ics.uci.edu/ml/machine-learning-databases/00296/dataset_diabetes.zip

In [ ]:
with zipfile.ZipFile("dataset_diabetes.zip", 'r') as zip_ref:
    zip_ref.extractall("diabetes_130")

# Data Wrangling

In [4]:
data = pd.read_csv("diabetes_130/dataset_diabetes/diabetic_data.csv",
                   sep=",",
                   index_col="encounter_id",
                   na_values="?")

/Users/nastassya.horlava/opt/anaconda3/envs/ehrapy_env/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3457: DtypeWarning: Columns (10) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


In [5]:
data.head()

patient_nbr             race  gender      age weight  \
encounter_id                                                         
2278392           8222157        Caucasian  Female   [0-10)    NaN   
149190           55629189        Caucasian  Female  [10-20)    NaN   
64410            86047875  AfricanAmerican  Female  [20-30)    NaN   
500364           82442376        Caucasian    Male  [30-40)    NaN   
16680            42519267        Caucasian    Male  [40-50)    NaN   

              admission_type_id  discharge_disposition_id  \
encounter_id                                                
2278392                       6                        25   
149190                        1                         1   
64410                         1                         1   
500364                        1                         1   
16680                         1                         1   

              admission_source_id  time_in_hospital payer_code  \
encounter_id                                                     
2278392                         1                 1        NaN   
149190                          7                 3        NaN   
64410                           7                 2        NaN   
500364                          7                 2        NaN   
16680                           7                 1        NaN   

                     medical_specialty  num_lab_procedures  num_procedures  \
encounter_id                                                                 
2278392       Pediatrics-Endocrinology                  41               0   
149190                             NaN                  59               0   
64410                              NaN                  11               5   
500364                             NaN                  44               1   
16680                              NaN                  51               0   

              num_medications  number_outpatient  number_emergency  \
encounter_id                                                         
2278392                     1                  0                 0   
149190                     18                  0                 0   
64410                      13                  2                 0   
500364                     16                  0                 0   
16680                       8                  0                 0   

              number_inpatient  diag_1  diag_2 diag_3  number_diagnoses  \
encounter_id                                                              
2278392                      0  250.83     NaN    NaN                 1   
149190                       0     276  250.01    255                 9   
64410                        1     648     250    V27                 6   
500364                       0       8  250.43    403                 7   
16680                        0     197     157    250                 5   

             max_glu_serum A1Cresult metformin repaglinide nateglinide  \
encounter_id                                                             
2278392               None      None        No          No          No   
149190                None      None        No          No          No   
64410                 None      None        No          No          No   
500364                None      None        No          No          No   
16680                 None      None        No          No          No   

             chlorpropamide glimepiride acetohexamide glipizide glyburide  \
encounter_id                                                                
2278392                  No          No            No        No        No   
149190                   No          No            No        No        No   
64410                    No          No            No    Steady        No   
500364                   No          No            No        No        No   
16680                    No          No            No    Steady        No   

         

In [6]:
data.describe(include="all")

patient_nbr       race  gender      age    weight  admission_type_id  \
count   1.017660e+05      99493  101766   101766      3197      101766.000000   
unique           NaN          5       3       10         9                NaN   
top              NaN  Caucasian  Female  [70-80)  [75-100)                NaN   
freq             NaN      76099   54708    26068      1336                NaN   
mean    5.433040e+07        NaN     NaN      NaN       NaN           2.024006   
std     3.869636e+07        NaN     NaN      NaN       NaN           1.445403   
min     1.350000e+02        NaN     NaN      NaN       NaN           1.000000   
25%     2.341322e+07        NaN     NaN      NaN       NaN           1.000000   
50%     4.550514e+07        NaN     NaN      NaN       NaN           1.000000   
75%     8.754595e+07        NaN     NaN      NaN       NaN           3.000000   
max     1.895026e+08        NaN     NaN      NaN       NaN           8.000000   

        discharge_disposition_id  admission_source_id  time_in_hospital  \
count              101766.000000        101766.000000     101766.000000   
unique                       NaN                  NaN               NaN   
top                          NaN                  NaN               NaN   
freq                         NaN                  NaN               NaN   
mean                    3.715642             5.754437          4.395987   
std                     5.280166             4.064081          2.985108   
min                     1.000000             1.000000          1.000000   
25%                     1.000000             1.000000          2.000000   
50%                     1.000000             7.000000          4.000000   
75%                     4.000000             7.000000          6.000000   
max                    28.000000            25.000000         14.000000   

       payer_code medical_specialty  num_lab_procedures  num_procedures  \
count       61510             51817       101766.000000   101766.000000   
unique         17                72                 NaN             NaN   
top            MC  InternalMedicine                 NaN             NaN   
freq        32439             14635                 NaN             NaN   
mean          NaN               NaN           43.095641        1.339730   
std           NaN               NaN           19.674362        1.705807   
min           NaN               NaN            1.000000        0.000000   
25%           NaN               NaN           31.000000        0.000000   
50%           NaN               NaN           44.000000        1.000000   
75%           NaN               NaN           57.000000        2.000000   
max           NaN               NaN          132.000000        6.000000   

        num_medications  number_outpatient  number_emergency  \
count     101766.000000      101766.000000     101766.000000   
unique              NaN                NaN               NaN   
top                 NaN                NaN               NaN   
freq                NaN                NaN               NaN   
mean          16.021844           0.369357          0.197836   
std            8.127566           1.267265          0.930472   
min            1.000000           0.000000          0.000000   
25%           10.000000           0.000000          0.000000   
50%           15.000000           0.000000          0.000000   
75%           20.000000           0.000000          0.000000   
max           81.000000          42.000000         76.000000   

        number_inpatient  diag_1  diag_2  diag_3  number_diagnoses  \
count      101766.000000  101745  101408  100343     101766.000000   
unique               NaN     716     748     789               NaN   
top                  NaN     428     276     250               NaN   
freq                 NaN    6862    6752   11555               NaN   
mean            0.635566     NaN     NaN     NaN          7.422607   
std             1.262863     NaN     NaN     

In [7]:
data.dtypes

patient_nbr                  int64
race                        object
gender                      object
age                         object
weight                      object
admission_type_id            int64
discharge_disposition_id     int64
admission_source_id          int64
time_in_hospital             int64
payer_code                  object
medical_specialty           object
num_lab_procedures           int64
num_procedures               int64
num_medications              int64
number_outpatient            int64
number_emergency             int64
number_inpatient             int64
diag_1                      object
diag_2                      object
diag_3                      object
number_diagnoses             int64
max_glu_serum               object
A1Cresult                   object
metformin                   object
repaglinide                 object
nateglinide                 object
chlorpropamide              object
glimepiride                 object
acetohexamide       

First, map gender, age and weigt to categories

I'll map age and weight to mean of their categories to have some value representation

### Categorical values mapping

first, we'll map gender to 2 categories

In [8]:
data.gender.unique()

array(['Female', 'Male', 'Unknown/Invalid'], dtype=object)

In [9]:
gender_mapping={"Female":0, "Male":1, 'Unknown/Invalid': None}
data['gender'] = data.gender.map(gender_mapping)
data.gender.unique()

array([ 0.,  1., nan])

Now its a bit tricky: we have numerical values (age and weight) as categories 
I'd propose to map it to mean of this value ranges to have some link to real values ranges

In [10]:
data.age.unique() #mean_aage columns and category for obs 

array(['[0-10)', '[10-20)', '[20-30)', '[30-40)', '[40-50)', '[50-60)',
       '[60-70)', '[70-80)', '[80-90)', '[90-100)'], dtype=object)

In [11]:
import numpy as np

data.age=data.apply(lambda x:np.mean([int(el) for el in x.age[1:-1].split('-')]), axis=1)
data.age.unique()

array([ 5., 15., 25., 35., 45., 55., 65., 75., 85., 95.])

In [12]:
data.weight.unique()

array([nan, '[75-100)', '[50-75)', '[0-25)', '[100-125)', '[25-50)',
       '[125-150)', '[175-200)', '[150-175)', '>200'], dtype=object)

with weight, we also have category '>200', i'd map it to 250 for now

In [13]:
import numpy as np
data["weight"]=data[data["weight"].notnull()].apply(lambda x: str(int(x.weight.split('>')[-1])+50) if (">" in x.weight) else x.weight, axis=1)
data["weight"]=data[data["weight"].notnull()].apply(lambda x: np.mean([int(el) for el in x.weight[1:-1].split('-')]) if "-" in x.weight else int(x.weight), axis=1)
data.weight.unique()


array([  nan,  87.5,  62.5,  12.5, 112.5,  37.5, 137.5, 187.5, 162.5,
       250. ])

Maping categorical values to ordinal

readmission attribute (outcome) has following values: 'NO' if no readmission, '>30' if readmitted to hospital after more than 30 days, '<30' if less than 30 days] 
in the paper https://downloads.hindawi.com/journals/bmri/2014/781670.pdf), they define 2 groups: “readmitted,” if the patient was readmitted within 30 days of discharge or “otherwise,” which covers both readmission after 30 days and no readmission at all. 

thus lets define two columns: first if the patint had early readmission(True) or late/no readmission 
column readmission type (0 for early, 1for late, 2 for No readmission) to have some ordering 
and remove original column

In [14]:
data.readmitted.unique()

array(['NO', '>30', '<30'], dtype=object)

In [15]:
readmission_map={"NO":False,">30":False,  "<30":True} #discard
data["early_readmission"]=data.readmitted.map(readmission_map)
print(data["early_readmission"].unique())

readmission_map={"NO":2.,">30":1, "<30":0.} #leave as strings 
data["readmission_class"]=data.readmitted.map(readmission_map)
print(data["readmission_class"].unique())

data.drop(columns={"readmitted"}, inplace=True)

[False  True]
[2. 1. 0.]


max_glu_serum: Indicates the range of the result or if the test was not taken. Values: “>200,” “>300,” “normal,” and “none” if not measured
mapping it to 0 for normal, >200 to 1, >300 to 2 to have ordering

In [19]:
data.max_glu_serum.unique()
max_glu_serum_map={"None":None, 'Norm':0, '>200':1, ">300":2}
data["max_glu_serum"]=data.max_glu_serum.map(max_glu_serum_map)
print(data["max_glu_serum"].unique())

[nan  2.  0.  1.]


Mapping categorical that represent boolenan to bool types

In [20]:
#Map diabetesMed to bool 
data.diabetesMed.unique()
diabetesMed_map={"No":False, 'Yes':True}
data["diabetesMed"]=data.diabetesMed.map(diabetesMed_map)
print(data["diabetesMed"].unique())

[False  True]


In [21]:
#Map change of medications to bool and rename it

print(data.change.unique())
data.change.unique()
medChange_map={"No":False, 'Ch':True}
data["change"]=data.change.map(medChange_map)

print(data["change"].unique())

['No' 'Ch']
[False  True]


Now, below cols indicate whether the drug was prescribed or there was a change in the dosage. Values: “up” if the dosage was increased during the encounter, “down” if the dosage was decreased, “steady” if the dosage did not change, and “no” if the drug was not prescribed
We'll map them to categories

In [22]:
cols=["metformin", "repaglinide", "nateglinide", "chlorpropamide", "glimepiride", 
      "acetohexamide", "glipizide", "glyburide", "tolbutamide", "pioglitazone", "rosiglitazone", 
      "acarbose", "miglitol", "troglitazone", "tolazamide", "examide", "citoglipton",
      "insulin", "glyburide-metformin", "glipizide-metformin", "glimepiride-pioglitazone", 
      "metformin-rosiglitazone", "metformin-pioglitazone"]
for col in cols:
    print(col, data[col].unique())

metformin ['No' 'Steady' 'Up' 'Down']
repaglinide ['No' 'Up' 'Steady' 'Down']
nateglinide ['No' 'Steady' 'Down' 'Up']
chlorpropamide ['No' 'Steady' 'Down' 'Up']
glimepiride ['No' 'Steady' 'Down' 'Up']
acetohexamide ['No' 'Steady']
glipizide ['No' 'Steady' 'Up' 'Down']
glyburide ['No' 'Steady' 'Up' 'Down']
tolbutamide ['No' 'Steady']
pioglitazone ['No' 'Steady' 'Up' 'Down']
rosiglitazone ['No' 'Steady' 'Up' 'Down']
acarbose ['No' 'Steady' 'Up' 'Down']
miglitol ['No' 'Steady' 'Down' 'Up']
troglitazone ['No' 'Steady']
tolazamide ['No' 'Steady' 'Up']
examide ['No']
citoglipton ['No']
insulin ['No' 'Up' 'Steady' 'Down']
glyburide-metformin ['No' 'Steady' 'Down' 'Up']
glipizide-metformin ['No' 'Steady']
glimepiride-pioglitazone ['No' 'Steady']
metformin-rosiglitazone ['No' 'Steady']
metformin-pioglitazone ['No' 'Steady']


discard coliumns that have one value everywhere
encode all other medications : {'No':0, 'Down':1, 'Steady':2, 'Up':3}

In [23]:
med_regulation_map={'No':0, 'Down':1, 'Steady':2, 'Up':3}  
for col in cols: 
    if len(data[col].unique())==1:
        data=data.drop(columns=[col])
        continue
    else:
        data[col]=data[col].map(med_regulation_map)

In [24]:
for col in cols:
    if col in data.columns:
        print(col, data[col].unique())

metformin [0 2 3 1]
repaglinide [0 3 2 1]
nateglinide [0 2 1 3]
chlorpropamide [0 2 1 3]
glimepiride [0 2 1 3]
acetohexamide [0 2]
glipizide [0 2 3 1]
glyburide [0 2 3 1]
tolbutamide [0 2]
pioglitazone [0 2 3 1]
rosiglitazone [0 2 3 1]
acarbose [0 2 3 1]
miglitol [0 2 1 3]
troglitazone [0 2]
tolazamide [0 2 3]
insulin [0 3 2 1]
glyburide-metformin [0 2 1 3]
glipizide-metformin [0 2]
glimepiride-pioglitazone [0 2]
metformin-rosiglitazone [0 2]
metformin-pioglitazone [0 2]


### Disease categories

Map disease types to categories to have categorical and not numerical data in disease column 
map according to table 2 of paper https://downloads.hindawi.com/journals/bmri/2014/781670.pdf
we'll do it for primary and secondary diadnosis separately


In [ ]:
cols=["diag_1", "diag_2", "diag_3"]

for col in cols
    print(data[col].unique())

In [64]:
"V12.32".split(".")[0].isnumeric()

False

In [69]:
str_vals = [chr(i) for i in range(ord('E'),ord('V')+1)]
# isfloat(data.iloc[0]["diag_1"])
data2=data[(data["diag_2"].notnull()) & (data["diag_2"].str.contains('E'))]
data2.head()


#TODO: multiple mock columns: 
#contain ., start with str_vals, normal numeric value
#then match each of this columns
# data[data["diag_2"].notnull() & data["diag_2"].str.split(".")[0].isnumeric(), axis=1)


KeyError: 0

In [27]:
cols=["diag_1", "diag_2", "diag_3"]
str_vals = [chr(i) for i in range(ord('E'),ord('V')+1)]
for col in cols:
    data[(data[col].notnull()) & (data[col].str.isnumeric()) & ]

['250.83' '276' '648' '8' '197' '414' '428' '398' '434' '250.7']
[nan '250.01' '250' '250.43' '157' '411' '492' '427' '198' '403']
[nan '255' 'V27' '403' '250' 'V45' '38' '486' '996' '197']


## Map all "unknown" and "non mapped" subtypes to None

Some columns are categorical but have multiple codes meaning "Unknown", lets map them all to None
Data for that is taken from "IDs_mappings"

In [16]:
print(data.admission_type_id.unique())
data.admission_type_id=data.apply(lambda x:None if x.admission_type_id in [5,6,8] else x.admission_type_id, axis=1)
print(data.admission_type_id.unique())


[6 1 2 3 4 5 8 7]
[nan  1.  2.  3.  4.  7.]


In [17]:
print(data.discharge_disposition_id.unique())
data.discharge_disposition_id=data.apply(lambda x:None if x.discharge_disposition_id in [18, 25, 26] else x.discharge_disposition_id, axis=1)
print(data.discharge_disposition_id.unique())

[25  1  3  6  2  5 11  7 10  4 14 18  8 13 12 16 17 22 23  9 20 15 24 28
 19 27]
[nan  1.  3.  6.  2.  5. 11.  7. 10.  4. 14.  8. 13. 12. 16. 17. 22. 23.
  9. 20. 15. 24. 28. 19. 27.]


In [18]:
print(data.admission_source_id.unique())
data.admission_source_id=data.apply(lambda x:None if x.admission_source_id in [9, 15,17,20, 21] else x.admission_source_id, axis=1)
print(data.admission_source_id.unique())

[ 1  7  2  4  5  6 20  3 17  8  9 14 10 22 11 25 13]
[ 1.  7.  2.  4.  5.  6. nan  3.  8. 14. 10. 22. 11. 25. 13.]


## Rename columns

Rename some columns for better understanding of the data

In [ ]:
#rename time column to be more decriptive
data=data.rename(columns={"time_in_hospital":"time_in_hospital_days", "change":"change_of_meds"})

cols=["number_outpatient", "number_emergency", "number_inpatient"]
for el in cols:
    data=data.rename(columns={el:f"{el}_visits"})

In [ ]:
data.describe(include="all")

### test 

In [ ]:
import ehrapy.api as ep
adata = ep.dt.mimic_2(encoded=False)

In [ ]:
adata.var_names


In [ ]:
adata.X[0]

In [ ]:
import pandas as pd
data_pd=pd.DataFrame(adata.X, columns=adata.var_names)

In [ ]:
data_pd.head()

In [ ]:
adata